compiler/rustc_mir/src/transform/const_prop.rs

```diff
@@ -476,6 +478,19 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             Ok(op) => Some(op),
             Err(error) => {
                 let tcx = self.ecx.tcx.at(c.span);
+                if error.kind().is_spurious() {
+                    // Spurious errors can't be ignored since otherwise the amount of available
+                    // memory influences the result of optimization and the build. The error
+                    // doesn't need to be fatal since no code will actually be generated anyways.
+                    self.ecx
+                        .tcx
+                        .tcx
+                        .sess
+                        .struct_err("memory exhausted during optimization")
+                        .help("try increasing the amount of memory available to the compiler")
+                        .emit();
```

Comment:
This won't hold up well when other errors are marked as "spurious".

The code below already has that `lint_only` path, can that just do `lint_only && !error.kind().is_hard_err()`?

Reply:
That would sidestep the `needs_subst` test for hard errors. I have no idea what consequences that has... @oli-obk?^^

Reply:
That will affect other code though, so it is probably better to first land another PR that uses `is_hard_err` here as well. It might need a crater run.

Reply:
Hm no actually this can't work, we can't hard error for promoteds in `const`/`static` bodies (that's the messy last bit of promotion heritage that is basically impossible to clean up).

But... do we even get here for `const`/`static` bodies? can we avoid running this optimization for them? Everything else should be fine with hard errors here I think (but I don't understand that `need_subst` stuff).

Reply:
Okay, here's my proposal:
replace `lint_only` by `lint_only && !matches(error.kind(), InterpError::ResourceExhaustion(ResourceExhaustionInfo::MemoryExhausted))`, and add a comment like `FIXME: use error.kind().is_hard_err(), but beware of backwards compatibility and interactions with promoteds`. Hopefully @oli-obk or me can pick this up at some point and take care of the FIXME.